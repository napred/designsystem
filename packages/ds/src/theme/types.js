// @flow

export type FontPalette = {
  default: string,
  [extra: string]: string,
};

export type ThemeSettings = {
  borderRadiuses: Array<string | number>,
  breakpoints: Array<string>,
  colors: { [name: string]: string },
  containerSizes: Array<number | string>,
  fontFamilies?: FontPalette,
  fontSizes: Array<number>,
  gutters: Array<number>,
  importFonts?: Array<string>,
  spacing: Array<number>,
};

export type Theme = {
  get(attributeName: string): any,
};
