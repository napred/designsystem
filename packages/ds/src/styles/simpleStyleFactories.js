// @flow

import convertUnit from '../utilities/convertUnit';
import formatStyleResult from './formatStyleResult';
import type { Styler } from './types';
import type { System } from '../system';

export function createStyle<T: Object>(
  propName: string,
  cssAttribute: string | Array<string>,
  formatter: (val: any) => string,
  defaultValue?: any,
): Styler<T> {
  const cssAttributes = Array.isArray(cssAttribute) ? cssAttribute : [cssAttribute];

  return {
    apply: (props: Object, { viewport: bp }: System) => {
      const propValue = props[propName] == null ? defaultValue : props[propName];

      if (propValue == null) {
        return formatStyleResult(propValue, cssAttributes, formatter);
      }

      if (Array.isArray(propValue)) {
        const bpPropValue = propValue[bp];
        const lastBpPropValue = propValue[propValue.length - 1];

        if (bpPropValue != null) {
          return formatStyleResult(bpPropValue, cssAttributes, formatter);
        }

        // if propValue for bp does not exist
        // look if we have defaultValue
        // if we dont have defaultvalue, return empty result
        // if defaultValue is scalar return it
        // if defaultValue is responsive (array) pick the longer array and return last value
        if (Array.isArray(defaultValue)) {
          const bpDefaultValue = defaultValue[bp];

          if (bpDefaultValue != null) {
            return formatStyleResult(bpDefaultValue, cssAttributes, formatter);
          }

          const arr = propValue.length >= defaultValue.length ? propValue : defaultValue;
          const val = arr[arr.length - 1];

          return formatStyleResult(val, cssAttributes, formatter);
        } else if (defaultValue != null) {
          return formatStyleResult(defaultValue, cssAttributes, formatter);
        }

        // return the last breakpoint value from an array if current viewport index
        // exceeds the range
        return formatStyleResult(lastBpPropValue, cssAttributes, formatter);
      }

      return formatStyleResult(propValue, cssAttributes, formatter);
    },
    propNames: [propName],
    stripProps: [propName],
  };
}

export function createNumericStyle<T: Object>(
  propName: string,
  cssAttribute: string | Array<string>,
  defaultValue?: number | string | Array<number | string> | void,
): Styler<T> {
  return createStyle(propName, cssAttribute, convertUnit, defaultValue);
}

export function createStringStyle<T: Object>(
  propName: string,
  cssAttribute: string | Array<string>,
  defaultValue?: string | Array<string> | void,
): Styler<T> {
  return createStyle(propName, cssAttribute, val => val, defaultValue);
}
