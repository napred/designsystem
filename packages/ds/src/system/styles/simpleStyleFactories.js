// @flow

import type { SystemComponentProps, StyleFn } from '../types';
import convertUnit from '../../utilities/convertUnit';
import formatStyleResult from './formatStyleResult';

export function createStyle<T: Object>(
  propName: string,
  cssAttribute: string | Array<string>,
  formatter: (val: any) => string,
  defaultValue?: any,
): StyleFn<T> {
  const cssAttributes = Array.isArray(cssAttribute) ? cssAttribute : [cssAttribute];

  const styler = (props: SystemComponentProps) => {
    const propValue = props[propName] == null ? defaultValue : props[propName];

    if (propValue == null) {
      return formatStyleResult(propValue, cssAttributes, formatter);
    }

    const bp = props.viewport;

    if (Array.isArray(propValue)) {
      const bpPropValue = propValue[bp];
      const lastBpPropValue = propValue[propValue.length - 1];

      if (bpPropValue != null) {
        return formatStyleResult(bpPropValue, cssAttributes, formatter);
      }

      // if propValue for bp does not exist
      // look if we have defaultValue
      // if we dont have defaultvalue, return empty result
      // if defaultValue is scalar return it
      // if defaultValue is responsive (array) pick the longer array and return last value
      if (Array.isArray(defaultValue)) {
        const bpDefaultValue = defaultValue[bp];

        if (bpDefaultValue != null) {
          return formatStyleResult(bpDefaultValue, cssAttributes, formatter);
        }

        const arr = propValue.length >= defaultValue.length ? propValue : defaultValue;
        const val = arr[arr.length - 1];

        return formatStyleResult(val, cssAttributes, formatter);
      } else if (defaultValue != null) {
        return formatStyleResult(defaultValue, cssAttributes, formatter);
      }

      // return the last breakpoint value from an array if current viewport index
      // exceeds the range
      return formatStyleResult(lastBpPropValue, cssAttributes, formatter);
    }

    return formatStyleResult(propValue, cssAttributes, formatter);
  };

  styler.propNames = [propName];

  return (styler: any);
}

export function createNumericStyle<T: Object>(
  propName: string,
  cssAttribute: string | Array<string>,
  defaultValue?: number | string | Array<number | string> | void,
): StyleFn<T> {
  return createStyle(propName, cssAttribute, convertUnit, defaultValue);
}

export function createStringStyle<T: Object>(
  propName: string,
  cssAttribute: string | Array<string>,
  defaultValue?: string | Array<string> | void,
): StyleFn<T> {
  return createStyle(propName, cssAttribute, val => val, defaultValue);
}
