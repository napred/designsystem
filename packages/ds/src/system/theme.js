// @flow

type FontPalette = {
  default: string,
  [extra: string]: string,
};

type ThemeSettings = {
  borderRadiuses: Array<string | number>,
  breakpoints: Array<string>,
  colors: { [name: string]: string },
  containerSizes: Array<number | string>,
  fontFamilies?: FontPalette,
  fontSizes: Array<number>,
  gutters: Array<number>,
  importFonts?: Array<string>,
  spacing: Array<number>,
};

export default class Theme {
  borderRadiuses: Array<string | number>;
  breakpoints: Array<string>;
  colors: { [name: string]: string };
  containerSizes: Array<number | string>;
  fontFamilies: FontPalette;
  fontSizes: Array<number>;
  gutters: Array<number>;
  importFonts: Array<string>;
  spacing: Array<number>;

  constructor(settings: ThemeSettings) {
    this.borderRadiuses = settings.borderRadiuses;
    this.breakpoints = settings.breakpoints;
    this.colors = settings.colors;
    this.containerSizes = settings.containerSizes;
    this.fontFamilies = settings.fontFamilies || {
      default: 'inherit', // inherit browser defaults
    };
    this.fontSizes = settings.fontSizes;
    this.gutters = settings.gutters;
    this.importFonts = settings.importFonts || [];
    this.spacing = settings.spacing;
  }

  color(name: string, defaultColor?: string): string {
    return (
      this.colors[name] ||
      (defaultColor ? this.colors[defaultColor] || defaultColor : this.colors.grey)
    );
  }

  containerSize(breakpoint: number): string | number {
    return this.containerSizes[breakpoint] || '100%';
  }

  get(attributeName: string): Array<string | number> | { [key: string]: string | number } {
    // $FlowExpectError
    const attributeValues = this[attributeName];

    if (attributeValues == null) {
      throw new Error(`There is no ${attributeName} defined in theme`);
    }

    return attributeValues;
  }

  gutter(breakpoint: number): number {
    return this.gutters[breakpoint] || 0;
  }

  mediaQueries(): Array<string> {
    return this.breakpoints;
  }

  space(scale: number): number {
    if (typeof this.spacing === 'number') {
      return this.spacing;
    }
    console.log('bla');
    return typeof this.spacing[scale] !== 'undefined'
      ? this.spacing[scale]
      : this.spacing[this.spacing.length - 1];
  }

  fontFamily(variant: string): string {
    return this.fontFamilies[variant] || this.fontFamilies.default;
  }

  fontSize(scale: number): number {
    if (typeof this.fontSizes === 'number') {
      return this.fontSizes;
    }
    return typeof this.fontSizes[scale] !== 'undefined'
      ? this.fontSizes[scale]
      : this.fontSizes[this.fontSizes.length - 1];
  }

  fonts(): string {
    const imports = this.importFonts.join(';');

    if (imports !== '') {
      return `${imports};`;
    }

    return '';
  }
}
