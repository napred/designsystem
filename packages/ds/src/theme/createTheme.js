// @flow

import type { ThemeSettings, Theme } from './types';

export default function createTheme(settings: ThemeSettings): Theme {
  return {
    get: (attributeName: string, defaultValue?: any): any => {
      const attributeValues = settings[attributeName];

      if (attributeValues == null) {
        if (defaultValue == null) {
          throw new Error(`There is no ${attributeName} defined in theme`);
        } else {
          return defaultValue;
        }
      } else if (
        typeof attributeValues === 'object' &&
        !Array.isArray(attributeValues) &&
        defaultValue != null
      ) {
        return attributeValues[defaultValue] || defaultValue;
      }

      return attributeValues;
    },
  };
}
