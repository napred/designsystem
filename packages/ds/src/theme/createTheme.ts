import { ITheme } from '../types';
import { IThemeSettings } from './types';

export default function createTheme(settings: IThemeSettings): ITheme {
  return {
    color(valueOrName, defaultValueOrName) {
      const color = settings.colors[valueOrName];

      if (color == null) {
        if (defaultValueOrName != null) {
          return settings.colors[defaultValueOrName] || defaultValueOrName;
        }

        return valueOrName;
      }

      return color;
    },
    get(attributeName, defaultValue) {
      const attributeValues = settings[attributeName];

      if (attributeValues == null) {
        if (defaultValue == null) {
          throw new Error(`There is no ${attributeName} defined in theme`);
        } else {
          return defaultValue;
        }
      } else if (
        typeof attributeValues === 'object' &&
        !Array.isArray(attributeValues) &&
        defaultValue != null
      ) {
        return attributeValues[defaultValue] || defaultValue;
      }

      return attributeValues;
    },

    getResponsiveValue(attributeName, viewport, defaultValue) {
      const attributeValues = settings[attributeName];

      if (attributeValues == null) {
        if (defaultValue == null) {
          throw new Error(`There is no ${attributeName} defined in theme`);
        } else {
          return defaultValue;
        }
      } else if (
        Array.isArray(attributeValues)
      ) {
        return attributeValues[viewport] || attributeValues[attributeValues.length - 1];
      }

      return attributeValues;
    },
  };
}
