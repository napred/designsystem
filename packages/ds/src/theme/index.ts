export * from './types';
export { default as createTheme } from './createTheme';
