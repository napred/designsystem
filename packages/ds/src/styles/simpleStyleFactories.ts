import { ISystem } from '../system';
import { IStyler } from '../types';
import { arrayize, convertUnit, getResponsiveValue } from '../utilities';
import formatStyleResult from './formatStyleResult';

export function createSimpleStyle<TProps extends object>(
  propName: keyof TProps,
  cssAttribute: string | string[],
  formatter: (val: any) => string | number,
  defaultValue?: string | number | null | undefined | Array<string | number | null | undefined>,
): IStyler<TProps> {
  const cssAttributes = Array.isArray(cssAttribute) ? cssAttribute : [cssAttribute];

  return {
    apply: (props: TProps, { viewport: bp }: ISystem) => {
      const propValue = arrayize<string | number | null | undefined>(props[propName] as any);
      const bpValue = getResponsiveValue(bp, propValue, defaultValue);

      return formatStyleResult(bpValue, cssAttributes, formatter);
    },
    propNames: [propName],
    stripProps: [propName],
  };
}

export function createNumericStyle<TProps extends object>(
  propName: keyof TProps,
  cssAttribute: string | string[],
  defaultValue?: string | number | null | undefined | Array<string | number | null | undefined>,
): IStyler<TProps> {
  return createSimpleStyle(propName, cssAttribute, convertUnit, defaultValue);
}

export function createStringStyle<TProps extends object>(
  propName: keyof TProps,
  cssAttribute: string | string[],
  defaultValue?: string | number | null | undefined | Array<string | number | null | undefined>,
): IStyler<TProps> {
  return createSimpleStyle(propName, cssAttribute, val => val, defaultValue);
}
