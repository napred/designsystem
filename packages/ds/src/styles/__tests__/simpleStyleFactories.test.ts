import theme from '../../defaultTheme';
import { createNumericStyle, createSimpleStyle, createStringStyle } from '../simpleStyleFactories';

describe('simple style factories', () => {
  describe('createStyler', () => {
    it('takes the value from default responsive value if is not specified in prop (with default)', () => {
      const style = createSimpleStyle<{ test?: any }>('test', 'test', val => val, ['1', '2']);

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
    });

    it('takes last value from responsive value if breakpoint exceeds the range (with default responsive)', () => {
      const style = createSimpleStyle<{ test?: any }>('test', 'test', val => val, ['1', '2']);

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();

      expect(style.apply({ test: ['1'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({ test: ['1', '3'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 2 })).toMatchSnapshot();
    });

    it('takes last value from responsive value if breakpoint exceeds the range (with default simple)', () => {
      const style = createSimpleStyle<{ test?: any }>('test', 'test', val => val, '1');

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();

      expect(style.apply({ test: ['1'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({ test: ['1', '3'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 2 })).toMatchSnapshot();
    });

    it('takes last value from responsive value if breakpoint exceeds the range (without default)', () => {
      const style = createSimpleStyle<{ test?: any }>('test', 'test', val => val);

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();

      expect(style.apply({ test: ['1'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({ test: ['1', '3'] }, { theme, viewport: 2 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 2 })).toMatchSnapshot();
    });
  });

  describe('createStringStyler', () => {
    it('works correctly (returns string)', () => {
      const style = createStringStyle<{ test?: any }>('test', 'test');

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();
    });
  });

  describe('createNumericStyler', () => {
    it('works correctly (returns unit)', () => {
      const style = createNumericStyle<{ test?: any }>('test', 'test');

      expect(style.apply({ test: 1 }, { theme, viewport: 0 })).toMatchSnapshot();
      expect(style.apply({}, { theme, viewport: 0 })).toMatchSnapshot();
    });
  });
});
