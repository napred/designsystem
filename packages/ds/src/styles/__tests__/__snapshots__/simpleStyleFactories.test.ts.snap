// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple style factories createNumericStyler works correctly (returns unit) 1`] = `
Object {
  "test": "1px",
}
`;

exports[`simple style factories createNumericStyler works correctly (returns unit) 2`] = `Object {}`;

exports[`simple style factories createStringStyler works correctly (returns string) 1`] = `
Object {
  "test": 1,
}
`;

exports[`simple style factories createStringStyler works correctly (returns string) 2`] = `Object {}`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default responsive) 1`] = `
Object {
  "test": 1,
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default responsive) 2`] = `
Object {
  "test": "1",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default responsive) 3`] = `
Object {
  "test": "2",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default responsive) 4`] = `
Object {
  "test": "3",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default responsive) 5`] = `
Object {
  "test": "2",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default simple) 1`] = `
Object {
  "test": 1,
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default simple) 2`] = `
Object {
  "test": "1",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default simple) 3`] = `
Object {
  "test": "1",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default simple) 4`] = `
Object {
  "test": "3",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (with default simple) 5`] = `
Object {
  "test": "1",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (without default) 1`] = `
Object {
  "test": 1,
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (without default) 2`] = `Object {}`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (without default) 3`] = `
Object {
  "test": "1",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (without default) 4`] = `
Object {
  "test": "3",
}
`;

exports[`simple style factories createStyler takes last value from responsive value if breakpoint exceeds the range (without default) 5`] = `Object {}`;

exports[`simple style factories createStyler takes the value from default responsive value if is not specified in prop (with default) 1`] = `
Object {
  "test": 1,
}
`;
